File: lib/src/features/monitoring/data/transport/ble_sensor_transport.dart
import 'dart:async';

import 'package:flutter_blue_plus/flutter_blue_plus.dart';

import '../../domain/sensor_reading.dart';
import '../reading_parser.dart';
import 'ble_permissions.dart';
import 'sensor_transport.dart';

/// BLE implementation backed by `flutter_blue_plus`.
///
/// Defaults to the Nordic UART Service, which is what ESP32 "BLE serial"
/// sketches expose. Override the UUIDs if your firmware uses its own.
class BleSensorTransport implements SensorTransport {
  BleSensorTransport({
    Guid? serviceUuid,
    Guid? notifyCharacteristicUuid,
  })  : serviceUuid = serviceUuid ?? nordicUartService,
        notifyCharacteristicUuid =
            notifyCharacteristicUuid ?? nordicUartTxCharacteristic {
    _adapterSubscription = FlutterBluePlus.adapterState.listen((state) {
      if (state != BluetoothAdapterState.on && !_statusController.isClosed) {
        _emit(
          const TransportStatus(
            state: TransportState.unavailable,
            message: 'Turn on Bluetooth to connect.',
          ),
        );
      } else if (state == BluetoothAdapterState.on &&
          _status.state == TransportState.unavailable) {
        _emit(const TransportStatus.disconnected());
      }
    });

    _scanSubscription = FlutterBluePlus.scanResults.listen((results) {
      _devices
        ..clear()
        ..addAll(
          results.map(
            (r) => DiscoveredDevice(
              id: r.device.remoteId.str,
              name: r.device.platformName.isNotEmpty
                  ? r.device.platformName
                  : r.advertisementData.advName,
              rssi: r.rssi,
            ),
          ),
        );
      if (!_devicesController.isClosed) {
        _devicesController.add(List<DiscoveredDevice>.unmodifiable(_devices));
      }
    });
  }

  /// Nordic UART Service — the de-facto standard for serial-over-BLE.
  static final Guid nordicUartService =
      Guid('6e400001-b5a3-f393-e0a9-e50e24dcca9e');

  /// Device to app characteristic (notify).
  static final Guid nordicUartTxCharacteristic =
      Guid('6e400003-b5a3-f393-e0a9-e50e24dcca9e');

  final Guid serviceUuid;
  final Guid notifyCharacteristicUuid;

  final StreamController<TransportStatus> _statusController =
      StreamController<TransportStatus>.broadcast();
  final StreamController<SensorReading> _readingsController =
      StreamController<SensorReading>.broadcast();
  final StreamController<List<DiscoveredDevice>> _devicesController =
      StreamController<List<DiscoveredDevice>>.broadcast();

  final List<DiscoveredDevice> _devices = <DiscoveredDevice>[];
  final PacketBuffer _packets = PacketBuffer();

  StreamSubscription<List<ScanResult>>? _scanSubscription;
  StreamSubscription<BluetoothAdapterState>? _adapterSubscription;
  StreamSubscription<BluetoothConnectionState>? _connectionSubscription;
  StreamSubscription<List<int>>? _valueSubscription;

  BluetoothDevice? _device;
  DiscoveredDevice? _descriptor;
  TransportStatus _status = const TransportStatus.disconnected();

  @override
  Stream<TransportStatus> get status async* {
    yield _status;
    yield* _statusController.stream;
  }

  @override
  Stream<SensorReading> get readings => _readingsController.stream;

  @override
  Stream<List<DiscoveredDevice>> get discovered async* {
    yield List<DiscoveredDevice>.unmodifiable(_devices);
    yield* _devicesController.stream;
  }

  @override
  Future<void> startScan({
    Duration timeout = const Duration(seconds: 10),
  }) async {
    if (!await FlutterBluePlus.isSupported) {
      _emit(
        const TransportStatus(
          state: TransportState.unavailable,
          message: 'This device has no Bluetooth LE radio.',
        ),
      );
      return;
    }

    // Asked here rather than in the controller so that non-BLE transports
    // never trigger a Bluetooth permission prompt.
    if (!await BlePermissions.ensureGranted()) {
      _emit(
        const TransportStatus(
          state: TransportState.unavailable,
          message: 'Bluetooth permission denied.',
        ),
      );
      return;
    }

    _devices.clear();
    _devicesController.add(const <DiscoveredDevice>[]);
    _emit(const TransportStatus(state: TransportState.scanning));

    try {
      // Scans for everything rather than filtering on the service UUID —
      // many dev boards do not advertise their service in the scan record.
      await FlutterBluePlus.startScan(timeout: timeout);
      await FlutterBluePlus.isScanning.where((scanning) => !scanning).first;
      if (_status.state == TransportState.scanning) {
        _emit(const TransportStatus.disconnected());
      }
    } on Exception catch (error) {
      _emit(TransportStatus(state: TransportState.error, message: '$error'));
    }
  }

  @override
  Future<void> stopScan() async {
    await FlutterBluePlus.stopScan();
    if (_status.state == TransportState.scanning) {
      _emit(const TransportStatus.disconnected());
    }
  }

  @override
  Future<void> connect(DiscoveredDevice device) async {
    await stopScan();
    await _teardownConnection();

    _descriptor = device;
    _emit(TransportStatus(state: TransportState.connecting, device: device));

    final target = BluetoothDevice.fromId(device.id);
    _device = target;

    try {
      // flutter_blue_plus v2 requires declaring a license tier. nonprofit
      // covers personal, nonprofit and educational use; commercial use needs
      // a paid licence from the package author.
      await target.connect(
        license: License.nonprofit,
        timeout: const Duration(seconds: 15),
      );

      _connectionSubscription = target.connectionState.listen((state) {
        if (state == BluetoothConnectionState.disconnected) {
          _packets.clear();
          _emit(const TransportStatus.disconnected());
        }
      });

      final characteristic = await _findNotifyCharacteristic(target);
      if (characteristic == null) {
        await target.disconnect();
        _emit(
          TransportStatus(
            state: TransportState.error,
            device: device,
            message: 'No notify characteristic found on this device.',
          ),
        );
        return;
      }

      _packets.clear();
      _valueSubscription = characteristic.onValueReceived.listen(_onBytes);
      await characteristic.setNotifyValue(true);

      _emit(TransportStatus(state: TransportState.connected, device: device));
    } on Exception catch (error) {
      _emit(
        TransportStatus(
          state: TransportState.error,
          device: device,
          message: '$error',
        ),
      );
    }
  }

  @override
  Future<void> disconnect() async {
    await _teardownConnection();
    _emit(const TransportStatus.disconnected());
  }

  @override
  Future<void> dispose() async {
    await _teardownConnection();
    await _scanSubscription?.cancel();
    await _adapterSubscription?.cancel();
    await _statusController.close();
    await _readingsController.close();
    await _devicesController.close();
  }

  Future<BluetoothCharacteristic?> _findNotifyCharacteristic(
    BluetoothDevice device,
  ) async {
    final services = await device.discoverServices();

    for (final service in services) {
      if (service.uuid != serviceUuid) continue;
      for (final characteristic in service.characteristics) {
        if (characteristic.uuid == notifyCharacteristicUuid) {
          return characteristic;
        }
      }
    }

    // Fall back to the first notifying characteristic on any service, so a
    // device with custom UUIDs still works without reconfiguring the app.
    for (final service in services) {
      for (final characteristic in service.characteristics) {
        final properties = characteristic.properties;
        if (properties.notify || properties.indicate) return characteristic;
      }
    }

    return null;
  }

  void _onBytes(List<int> bytes) {
    for (final line in _packets.add(bytes)) {
      final reading = ReadingParser.parse(
        line,
        deviceId: _descriptor?.id,
        deviceName: _descriptor?.displayName,
      );
      if (reading != null && !_readingsController.isClosed) {
        _readingsController.add(reading);
      }
    }
  }

  Future<void> _teardownConnection() async {
    await _valueSubscription?.cancel();
    _valueSubscription = null;
    await _connectionSubscription?.cancel();
    _connectionSubscription = null;
    _packets.clear();

    final device = _device;
    _device = null;
    if (device != null) {
      try {
        await device.disconnect();
      } on Exception {
        // Already gone — nothing useful to do.
      }
    }
  }

  void _emit(TransportStatus status) {
    _status = status;
    if (!_statusController.isClosed) _statusController.add(status);
  }
}

File: app/src/features/monitoring/data/transport/bleSensorTransport.ts
import { BleManager, State, type Device, type Subscription } from 'react-native-ble-plx';

import { base64ToBytes } from '@/core/utils/base64';

import type { SensorReading } from '../../domain/sensorReading';
import { PacketBuffer, parseReading } from '../readingParser';
import { ensureBlePermissions } from './blePermissions';
import {
  DISCONNECTED,
  Emitter,
  deviceDisplayName,
  type DiscoveredDevice,
  type SensorTransport,
  type TransportStatus,
  type Unsubscribe,
} from './sensorTransport';

/** Nordic UART Service — the de-facto standard for serial-over-BLE. */
export const NORDIC_UART_SERVICE = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';

/** Device to app characteristic (notify). */
export const NORDIC_UART_TX_CHARACTERISTIC = '6e400003-b5a3-f393-e0a9-e50e24dcca9e';

export interface BleTransportOptions {
  serviceUuid?: string;
  notifyCharacteristicUuid?: string;
  /** Injectable for tests; defaults to a fresh [BleManager]. */
  manager?: BleManager;
}

const sameUuid = (a: string, b: string) => a.toLowerCase() === b.toLowerCase();

/**
 * BLE implementation backed by `react-native-ble-plx`.
 *
 * Defaults to the Nordic UART Service, which is what ESP32 "BLE serial"
 * sketches expose. Override the UUIDs if your firmware uses its own.
 */
export class BleSensorTransport implements SensorTransport {
  readonly serviceUuid: string;
  readonly notifyCharacteristicUuid: string;

  private readonly manager: BleManager;
  private readonly statusEmitter = new Emitter<TransportStatus>();
  private readonly readingEmitter = new Emitter<SensorReading>();
  private readonly deviceEmitter = new Emitter<readonly DiscoveredDevice[]>();
  private readonly packets = new PacketBuffer();

  private devices: DiscoveredDevice[] = [];
  private currentStatus: TransportStatus = DISCONNECTED;
  private descriptor: DiscoveredDevice | null = null;
  private device: Device | null = null;

  private stateSubscription: Subscription | null = null;
  private disconnectSubscription: Subscription | null = null;
  private valueSubscription: Subscription | null = null;
  private scanTimer: ReturnType<typeof setTimeout> | null = null;
  private scanning = false;

  constructor({
    serviceUuid = NORDIC_UART_SERVICE,
    notifyCharacteristicUuid = NORDIC_UART_TX_CHARACTERISTIC,
    manager,
  }: BleTransportOptions = {}) {
    this.serviceUuid = serviceUuid;
    this.notifyCharacteristicUuid = notifyCharacteristicUuid;
    this.manager = manager ?? new BleManager();

    this.stateSubscription = this.manager.onStateChange((state) => {
      if (state !== State.PoweredOn) {
        this.emit({
          state: 'unavailable',
          device: null,
          message:
            state === State.Unsupported
              ? 'This device has no Bluetooth LE radio.'
              : 'Turn on Bluetooth to connect.',
        });
      } else if (this.currentStatus.state === 'unavailable') {
        this.emit(DISCONNECTED);
      }
    }, true);
  }

  get status(): TransportStatus {
    return this.currentStatus;
  }

  get discovered(): readonly DiscoveredDevice[] {
    return this.devices;
  }

  onStatus(listener: (status: TransportStatus) => void): Unsubscribe {
    listener(this.currentStatus);
    return this.statusEmitter.add(listener);
  }

  onReading(listener: (reading: SensorReading) => void): Unsubscribe {
    return this.readingEmitter.add(listener);
  }

  onDiscovered(listener: (devices: readonly DiscoveredDevice[]) => void): Unsubscribe {
    listener(this.devices);
    return this.deviceEmitter.add(listener);
  }

  async startScan(timeoutMs = 10_000): Promise<void> {
    const state = await this.manager.state();
    if (state === State.Unsupported) {
      this.emit({
        state: 'unavailable',
        device: null,
        message: 'This device has no Bluetooth LE radio.',
      });
      return;
    }

    // Asked here rather than in the store so that non-BLE transports never
    // trigger a Bluetooth permission prompt.
    if (!(await ensureBlePermissions())) {
      this.emit({ state: 'unavailable', device: null, message: 'Bluetooth permission denied.' });
      return;
    }

    if (state !== State.PoweredOn) {
      this.emit({ state: 'unavailable', device: null, message: 'Turn on Bluetooth to connect.' });
      return;
    }

    this.devices = [];
    this.deviceEmitter.emit([]);
    this.emit({ state: 'scanning', device: null, message: null });

    await new Promise<void>((resolve) => {
      this.scanning = true;

      const finish = async () => {
        if (!this.scanning) return;
        this.scanning = false;
        if (this.scanTimer) clearTimeout(this.scanTimer);
        this.scanTimer = null;
        try {
          await this.manager.stopDeviceScan();
        } catch {
          // The scan already stopped — nothing to do.
        }
        if (this.currentStatus.state === 'scanning') this.emit(DISCONNECTED);
        resolve();
      };

      // Scans for everything rather than filtering on the service UUID —
      // many dev boards do not advertise their service in the scan record.
      this.manager.startDeviceScan(null, { allowDuplicates: false }, (error, scanned) => {
        if (error) {
          this.emit({ state: 'error', device: null, message: error.message });
          void finish();
          return;
        }
        if (scanned) this.addDevice(scanned);
      });

      this.scanTimer = setTimeout(() => void finish(), timeoutMs);
      this.stopScanResolver = finish;
    });
  }

  private stopScanResolver: (() => Promise<void>) | null = null;

  async stopScan(): Promise<void> {
    const resolver = this.stopScanResolver;
    this.stopScanResolver = null;
    if (resolver) {
      await resolver();
    } else if (this.currentStatus.state === 'scanning') {
      this.emit(DISCONNECTED);
    }
  }

  async connect(device: DiscoveredDevice): Promise<void> {
    await this.stopScan();
    await this.teardownConnection();

    this.descriptor = device;
    this.emit({ state: 'connecting', device, message: null });

    try {
      // A larger MTU lets the ESP32 send a whole packet in one notification
      // instead of 20-byte fragments; it is a request, and the buffer copes
      // either way.
      const target = await this.manager.connectToDevice(device.id, {
        timeout: 15_000,
        requestMTU: 185,
      });
      this.device = target;

      this.disconnectSubscription = this.manager.onDeviceDisconnected(device.id, () => {
        this.packets.clear();
        this.emit(DISCONNECTED);
      });

      await target.discoverAllServicesAndCharacteristics();
      const characteristic = await this.findNotifyCharacteristic(target);
      if (characteristic == null) {
        await this.teardownConnection();
        this.emit({
          state: 'error',
          device,
          message: 'No notify characteristic found on this device.',
        });
        return;
      }

      this.packets.clear();
      this.valueSubscription = this.manager.monitorCharacteristicForDevice(
        device.id,
        characteristic.serviceUUID,
        characteristic.uuid,
        (error, value) => {
          if (error) {
            // A monitor error after a disconnect is expected noise; a live
            // one means the link is unusable.
            if (this.device != null) {
              this.emit({ state: 'error', device, message: error.message });
            }
            return;
          }
          if (value?.value) this.onBytes(base64ToBytes(value.value));
        },
      );

      this.emit({ state: 'connected', device, message: null });
    } catch (error) {
      this.emit({ state: 'error', device, message: describe(error) });
    }
  }

  async disconnect(): Promise<void> {
    await this.teardownConnection();
    this.emit(DISCONNECTED);
  }

  async dispose(): Promise<void> {
    await this.stopScan();
    await this.teardownConnection();
    this.stateSubscription?.remove();
    this.stateSubscription = null;
    this.statusEmitter.clear();
    this.readingEmitter.clear();
    this.deviceEmitter.clear();
    this.manager.destroy();
  }

  private addDevice(scanned: Device): void {
    const entry: DiscoveredDevice = {
      id: scanned.id,
      name: scanned.name ?? scanned.localName ?? '',
      rssi: scanned.rssi ?? 0,
    };
    const index = this.devices.findIndex((d) => d.id === entry.id);
    if (index >= 0) {
      this.devices[index] = entry;
    } else {
      this.devices.push(entry);
    }
    this.devices = [...this.devices];
    this.deviceEmitter.emit(this.devices);
  }

  private async findNotifyCharacteristic(device: Device) {
    const services = await device.services();

    for (const service of services) {
      if (!sameUuid(service.uuid, this.serviceUuid)) continue;
      for (const characteristic of await service.characteristics()) {
        if (sameUuid(characteristic.uuid, this.notifyCharacteristicUuid)) return characteristic;
      }
    }

    // Fall back to the first notifying characteristic on any service, so a
    // device with custom UUIDs still works without reconfiguring the app.
    for (const service of services) {
      for (const characteristic of await service.characteristics()) {
        if (characteristic.isNotifiable || characteristic.isIndicatable) return characteristic;
      }
    }

    return null;
  }

  private onBytes(bytes: Uint8Array): void {
    for (const line of this.packets.add(bytes)) {
      const reading = parseReading(line, {
        deviceId: this.descriptor?.id,
        deviceName: this.descriptor ? deviceDisplayName(this.descriptor) : null,
      });
      if (reading != null) this.readingEmitter.emit(reading);
    }
  }

  private async teardownConnection(): Promise<void> {
    this.valueSubscription?.remove();
    this.valueSubscription = null;
    this.disconnectSubscription?.remove();
    this.disconnectSubscription = null;
    this.packets.clear();

    const device = this.device;
    this.device = null;
    if (device != null) {
      try {
        await this.manager.cancelDeviceConnection(device.id);
      } catch {
        // Already gone — nothing useful to do.
      }
    }
  }

  private emit(status: TransportStatus): void {
    this.currentStatus = status;
    this.statusEmitter.emit(status);
  }
}

function describe(error: unknown): string {
  if (error instanceof Error) return error.message;
  return String(error);
}
